src/riak_kv_qry_buffers.erl

```diff
@@ -0,0 +1,815 @@
+          chunks_got = 0 :: integer(),
+          %% total chunks needed (== number of subqueries)
+          chunks_need :: integer(),
+
+          %% flipped to true when chunks_need == chunks_got
+          is_ready = false :: boolean(),
+
+          %% total records stored (when complete)
+          total_records = 0 :: non_neg_integer(),
+
+          %% %% iterator cache (need a working iterator support in eleveldb)
+          %% iter_cache = [] :: [cached_iter()],
+
+          %% total size on disk, for reporting
+          size = 0 :: non_neg_integer(),
+
+          %% last added chunk (when not complete) or queried
+          last_accessed = 0 :: erlang:timestamp(),
+
+          %% a proplist for all your options
+          options = [] :: qbuf_options(),
+
+          %% precomputed key field positions
+          key_field_positions :: [non_neg_integer()],
+
+          %% process waiting for this qbuf ready status
+          ready_waiting_pid :: pid() | undefined
+         }).
+
+-record(state, {
+          qbufs = [] :: [{qbuf_ref(), #qbuf{}}],
+          total_size = 0 :: non_neg_integer(),
+          %% no new queries; accumulation allowed
+          soft_watermark :: non_neg_integer(),
+          %% drop some tables now
+          hard_watermark :: non_neg_integer(),
+          %% drop incomplete query buffer after this long since last add_chunk
+          incomplete_qbuf_release_msec :: non_neg_integer(),
+          %% drop complete query buffers after this long since serving last query
+          qbuf_expire_msec :: non_neg_integer(),
+          %% dir containing qbuf ldb files
+          root_path :: string()
+         }).
+
+
+-spec start_link() -> {ok, pid()} | ignore | {error, term()}.
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+
+-spec init(proplists:proplist()) -> {ok, #state{}}.
+%% @private
+init(Options) ->
+    RootPath =
+        filename:join(
+          app_helper:get_prop_or_env(data_root, Options, eleveldb),  %% reuse eleveldb own root_path
+          "query_buffers"),
+    %% don't bother recovering any leftover tables
+    os:cmd(
+      fmt("rm -rf '~s'/*", [RootPath])),
+    State =
+        #state{soft_watermark =
+                   proplists:get_value(soft_watermark, Options, ?SOFT_WATERMARK),
+               hard_watermark =
+                   proplists:get_value(hard_watermark, Options, ?HARD_WATERMARK),
+               incomplete_qbuf_release_msec =
+                   proplists:get_value(incomplete_qbuf_release_msec, Options, ?INCOMPLETE_QBUF_RELEASE_MSEC),
+               qbuf_expire_msec =
+                   proplists:get_value(qbuf_expire_msec, Options, ?QBUF_EXPIRE_MSEC),
+               root_path =
+                   proplists:get_value(root_path, Options, RootPath)
+              },
+    Self = self(),
+    _TickerPid = spawn_link(fun() -> schedule_tick(Self) end),
+    {ok, State}.
+
+schedule_tick(Pid) ->
+    Pid ! tick,
+    timer:sleep(1000),
```

Comment:
Macroised (20000d0). As for parameters, they are now configurable in riak_kv.schema (0c15cd0).